src/service/xhr-impl.js

```diff
@@ -106,6 +117,18 @@ export class Xhr {
    * @private
    */
   fetch_(input, init) {
+    if (this.ampdoc_) {
+      // TODO(@jridgewell, #9048): We can alternatively _always_ wait on
+      // #whenFirstVisible.
+      if (!viewerForDoc(this.ampdoc_).hasBeenVisible()) {
+        dev().error('XHR', 'attempted to fetch %s before viewer was visible',
```

Comment:
Well, we don't want to overwhelm that service either. But basically, error transformer can only sample randomly regardless of error. So if we think this could have too big a rate, we should do it here.